ProcessLib/DeactivatedSubdomain.cpp

```diff
@@ -0,0 +1,164 @@
+#include "BaseLib/Error.h"
+
+#include "MeshLib/MeshEditing/DuplicateMeshComponents.h"
+#include "MeshLib/Mesh.h"
+#include "MeshLib/Elements/Element.h"
+#include "MeshLib/Node.h"
+
+namespace ProcessLib
+{
+const std::string DeactivatedSubdomain::name_of_paramater_of_zero =
+    "zero_for_element_deactivation_approach";
+
+DeactivetedSubdomainMesh::DeactivetedSubdomainMesh(
+    std::unique_ptr<MeshLib::Mesh> deactivated_subdomain_mesh_,
+    std::vector<MeshLib::Node*>&& inactive_nodes_)
+    : mesh(std::move(deactivated_subdomain_mesh_)),
+      inactive_nodes(std::move(inactive_nodes_))
+{
+}
+
+DeactivatedSubdomain::DeactivatedSubdomain(
+    std::unique_ptr<BaseLib::TimeInterval> time_interval_,
+    std::vector<int>&& materialIDs_,
+    std::vector<std::unique_ptr<DeactivetedSubdomainMesh>>&&
+        deactivated_sudomain_meshes_)
+    : time_interval(std::move(time_interval_)),
+      materialIDs(std::move(materialIDs_)),
+      deactivated_sudomain_meshes(std::move(deactivated_sudomain_meshes_))
+{
+}
+
+bool DeactivatedSubdomain::includesTimeOf(double const t) const
+{
+    return time_interval->contains(t);
+}
+
+std::unique_ptr<DeactivatedSubdomain const> createDeactivatedSubdomain(
+    BaseLib::ConfigTree const& config, MeshLib::Mesh const& mesh)
+{
+    //! \ogs_file_param{prj__process_variables__process_variable__deactivated_subdomains__deactivated_subdomain__time_interval}
+    config.peekConfigParameter<std::string>("time_interval");
+    auto time_interval = BaseLib::createTimeInterval(config);
+
+    std::vector<int> deactivated_subdomain_material_ids;
+    deactivated_subdomain_material_ids =
+        //! \ogs_file_param{prj__process_variables__process_variable__deactivated_subdomains__deactivated_subdomain__material_ids}
+        config.getConfigParameter<std::vector<int>>("material_ids",
+                                                    std::vector<int>{});
+
+    if (deactivated_subdomain_material_ids.empty())
+    {
+        OGS_FATAL(
+            "The material IDs of the deactivated subdomains are not given.");
+    }
+
+    std::sort(deactivated_subdomain_material_ids.begin(),
+              deactivated_subdomain_material_ids.end());
+
+    auto const* const material_ids = MeshLib::materialIDs(mesh);
+
+    std::vector<std::unique_ptr<DeactivetedSubdomainMesh>>
+        deactivated_sudomain_meshes;
+    deactivated_sudomain_meshes.reserve(
+        deactivated_subdomain_material_ids.size());
+
+    for (auto const ids : deactivated_subdomain_material_ids)
+    {
+        std::vector<MeshLib::Element*> deactivated_elements;
+        std::vector<MeshLib::Node*> deactivated_nodes;
+
+        // temporary vector to enhance node searching.
+        std::vector<bool> deactivation_flag_of_nodes(mesh.getNumberOfNodes(),
+                                                     false);
+
+        for (std::size_t i = 0; i < mesh.getNumberOfElements(); i++)
+        {
+            if (ids != (*material_ids)[i])
```

Comment:
Please add an `OGS_FATAL` if the `material_ids` are needed but not available. The pointer might be null.

Reply:
Moved the checking of material_ids form the ctor of Process to this function.